### PR TITLE
build_arg_string: Accept a dict containing configurable GMT parameters

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Fixes https://github.com/GenericMappingTools/pygmt/issues/2299

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by willschlitzer** [comment]
Should there be a check on keys passed to `confdict` to make sure that they are correct configuration names? Something like:
```
pygmt_configs = ["COLOR_BACKGROUND", "COLOR_FOREGROUND",  "COLOR_CPT", "COLOR_NAN", "COLOR_MODEL"...]
for key in confdict:
    if key not in pygmt_configs:
        raise GMTInvalidInput(f"{key} not a valid PyGMT configuration key.")
````

**Comment by seisman** [comment]
> Should there be a check on keys passed to `confdict` to make sure that they are correct configuration names? Something like:
> 
> ```
> pygmt_configs = ["COLOR_BACKGROUND", "COLOR_FOREGROUND",  "COLOR_CPT", "COLOR_NAN", "COLOR_MODEL"...]
> for key in confdict:
>     if key not in pygmt_configs:
>         raise GMTInvalidInput(f"{key} not a valid PyGMT configuration key.")
> ```

I don't think it's necessary, because `confdict` is only used internally by developers (https://github.com/GenericMappingTools/pygmt/issues/2299#issuecomment-1375008099), and developers should be responsible for passing the correct configuration names.

File: pygmt/helpers/utils.py [hunk 1]
```diff
@@ -119,16 +119,16 @@ def dummy_context(arg):
     yield arg
 
 
-def build_arg_string(kwdict, infile=None, outfile=None):
+def build_arg_string(kwdict, confdict=None, infile=None, outfile=None):
     r"""
-    Convert a dict and optional input/output files into a GMT argument string.
+    Convert keyword dicts and input/output files into a GMT argument string.
```
Comment (michaelgrund) [inline pygmt/helpers/utils.py:124]:
```suggestion
    Convert keyword dictionaries and input/output files into a GMT argument string.
```

File: pygmt/helpers/utils.py [hunk 2]
```diff
@@ -140,6 +140,8 @@ def build_arg_string(kwdict, infile=None, outfile=None):
     ----------
     kwdict : dict
         A dict containing parsed keyword arguments.
```
Comment (michaelgrund) [inline pygmt/helpers/utils.py:142]:
```suggestion
        A dictionary containing parsed keyword arguments.
```

File: pygmt/helpers/utils.py [hunk 3]
```diff
@@ -140,6 +140,8 @@ def build_arg_string(kwdict, infile=None, outfile=None):
...
+    confdict : dict
+        A dict containing GMT's configurations.
```
Comment (michaelgrund) [inline pygmt/helpers/utils.py:144]:
```suggestion
        A dictionary containing GMT's configurations.
```

**Comment by willschlitzer** [comment]
> > Should there be a check on keys passed to `confdict` to make sure that they are correct configuration names? Something like:
> > ```
> > pygmt_configs = ["COLOR_BACKGROUND", "COLOR_FOREGROUND",  "COLOR_CPT", "COLOR_NAN", "COLOR_MODEL"...]
> > for key in confdict:
> >     if key not in pygmt_configs:
> >         raise GMTInvalidInput(f"{key} not a valid PyGMT configuration key.")
> > ```
> 
> I don't think it's necessary, because `confdict` is only used internally by developers ([#2299 (comment)](https://github.com/GenericMappingTools/pygmt/issues/2299#issuecomment-1375008099)), and developers should be responsible for passing the correct configuration names.

I don't think I understand; isn't the use of `confdict` to allow users to set GMT parameters more easily?

**Comment by seisman** [comment]
> I don't think I understand; isn't the use of `confdict` to allow users to set GMT parameters more easily?

No, users should always use `pygmt.config` to set GMT parameters. 

When I try to implement the `Figure.timestamp()` method (https://github.com/GenericMappingTools/pygmt/pull/2208), I need to pass  some GMT parameters to the GMT C API, like:
```
    with Session() as lib:
        lib.call_module(
            module="plot",
            args=build_arg_string(kwdict)
            + f' --FONT_LOGO="{font}" --FORMAT_TIME_STAMP="{timefmt}"',
        )
```
I realize that using `build_arg_string(kwdict) + f' --FONT_LOGO="{font}" --FORMAT_TIME_STAMP="{timefmt}"'` is not convenient, so I propose to allow `build_arg_string` to accept configuration dicts. After this PR, the above code should be simplified to:
```
    with Session() as lib:
        lib.call_module(
            module="plot",
            args=build_arg_string(
                     kwdict, 
                     confdict={"FONT_LOGO": f"{font}", "FORMAT_TIME_STAMP": f"{timefmt}"},
            )
        )
```

**Comment by willschlitzer** [comment]
Okay, that makes sense.

File: pygmt/helpers/utils.py [hunk 4]
```diff
@@ -139,7 +140,9 @@ def build_arg_string(kwdict, infile=None, outfile=None):
     Parameters
     ----------
     kwdict : dict
-        A dict containing parsed keyword arguments.
+        A dictionary containing parsed keyword arguments.
+    confdict : dict
+        A dictionary containing GMT's configurations.
```
Comment (weiji14) [inline pygmt/helpers/utils.py:145]: Should we also state that this is optional?
```suggestion
        A dictionary containing configurable GMT parameters.
```

File: pygmt/helpers/utils.py [hunk 5]
```diff
@@ -149,8 +152,10 @@ def build_arg_string(kwdict, infile=None, outfile=None):
     -------
     args : str
         The space-delimited argument string with '-' inserted before each
-        keyword. The arguments are sorted alphabetically, with optional input
-        file at the beginning and optional output file at the end.
+        keyword, or '--' inserted before GMT's configuration key-value pairs.
```
Comment (weiji14) [inline pygmt/helpers/utils.py:155]:
```suggestion
        keyword, or '--' inserted before GMT configuration key-value pairs.
```

File: pygmt/helpers/utils.py [hunk 6]
```diff
@@ -149,8 +152,10 @@ def build_arg_string(kwdict, infile=None, outfile=None):
...
+        The keyword arguments are sorted alphabetically, followed by GMT's
```
Comment (weiji14) [inline pygmt/helpers/utils.py:156]:
```suggestion
        The keyword arguments are sorted alphabetically, followed by GMT
```

File: pygmt/helpers/utils.py (hunk 4 above)
Comment (seisman) [reply pygmt/helpers/utils.py]: The function definition says `confdict=None`. For me, it's clear that `confdict` is optional.

Comment (weiji14) [reply pygmt/helpers/utils.py]: Ok

**Review by weiji14 — APPROVED** [review]

**Review by willschlitzer — APPROVED** [review]

**Pull request merged**